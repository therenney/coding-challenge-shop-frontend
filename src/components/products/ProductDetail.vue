<template>
    <div>
        <div v-if="product">
            <h2>
                {{ product.name }}<br /><span class="uk-text-lead">{{ product.price | currency }}</span>
            </h2>
            <div class="uk-grid-small uk-margin-bottom" uk-grid>
                <div class="uk-width-1-1 uk-width-1-2@s uk-width-1-3@m">
                    <img
                        class="product-image"
                        :data-src="product.detailsPhotoUri"
                        width="100%"
                        height="150"
                        alt=""
                        uk-img
                    />
                </div>
                <div class="uk-width-1-1 uk-width-1-2@s uk-width-2-3@m">
                    <p>{{ product.description }}</p>
                </div>
            </div>
            <div class="uk-column-1-3@m">
                <dl class="uk-description-list">
                    <dt>Manufacturer</dt>
                    <dd>
                        {{ product.manufacturer }}
                    </dd>
                    <dt>Packaging</dt>
                    <dd>
                        {{ product.packaging }}
                    </dd>
                    <dt>Base Product Packaging</dt>
                    <dd>
                        {{ product.baseProductPackaging }}
                    </dd>
                    <dt>Base Product Unit</dt>
                    <dd>
                        {{ product.baseProductUnit }}
                    </dd>
                    <dt>Base Product Amount</dt>
                    <dd>
                        {{ product.baseProductAmount }}
                    </dd>
                    <dt>Base Product Quantity</dt>
                    <dd>
                        {{ product.baseProductQuantity }}
                    </dd>
                </dl>
            </div>
        </div>
        <div v-else="">
            <h2>
                Sorry, this product is currently not available!
            </h2>
        </div>
    </div>
</template>

<script>
export default {
    /**
     * Props definition
     * - products is required and needs to be an array
     */
    props: {
        product: {
            type: Object,
        },
    },

    /**
     * Filters - Add currency filter for Table
     * Should be made globally available but for this SPA i keep it in the component
     */
    filters: {
        currency: function(value, digits) {
            if (typeof value !== 'number') {
                return value;
            }

            const formatter = new Intl.NumberFormat('de-DE', {
                style: 'currency',
                currency: 'EUR',
                minimumFractionDigits: digits,
            });

            return formatter.format(value);
        },
    },
};
</script>
