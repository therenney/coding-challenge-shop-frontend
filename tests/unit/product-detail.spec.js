import { shallowMount } from '@vue/test-utils';
import ProductDetail from '@components/products/ProductDetail.vue';

const productData = {
    "id": "ABC1234567800",
    "gtin": "0544900006170400",
    "manufacturer": "Beverages Ltd.",
    "name": "Beverage 23, 6x 0.75 L",
    "description": "Lorem ipsum usu amet dicat nullam ea. Nec detracto lucilius democritum in, ne usu delenit offendit deterruisset. Recusabo iracundia molestiae ea pro, suas dicta nemore an cum, erat dolorum nonummy mel ea. Iisque labores liberavisse in mei, dico laoreet elaboraret nam et, iudico verterem platonem est an. Te usu paulo vidisse epicuri, facilis mentitum liberavisse vel ut, movet iriure invidunt ut quo. Ad melius mnesarchum scribentur eum, mel at mundi impetus utroque.",
    "packaging": "CA",
    "baseProductPackaging": "BO",
    "baseProductUnit": "LT",
    "baseProductAmount": 0.75,
    "baseProductQuantity": 6,
    "price": 459,
    "logoUri": "https://dummyimage.com/200x200/eeeeee/2fd6ac.png&text=Beverages+Ltd.",
    "listingPhotoUri": "https://dummyimage.com/300x200/eeeeee/2fd6ac.png&text=Beverage+23,+6x0.75",
    "detailsPhotoUri": "https://dummyimage.com/600x400/eeeeee/2fd6ac.png&text=Beverage+23,+6x0.75"
};

describe('ProductDetail', () => {

    describe('Given there is no product data', () => {

        describe('When the ProductDetail component is rendered', () => {

            it('should render a no data message', () => {

                const wrapper = shallowMount(ProductDetail, {
                    propsData: {
                        product: {},
                    },
                });

                // Expect that a product-error is rendered
                expect(wrapper.find('.product-error').exists()).toBe(true);

                // Expect the message to equal a given string
                expect(wrapper.find('h2').text()).toMatch('Sorry, this product is currently not available!');
            });
        });
    });


    describe('Given there is product data', () => {

        describe('When the ProductDetail component is rendered', () => {

            it('should render a view with the product details', () => {

                const wrapper = shallowMount(ProductDetail, {
                    propsData: {
                        product: productData,
                    },
                    stubs: {
                        'router-link': true,
                    },
                });

                // Expect that the table is rendered
                expect(wrapper.find('.product-detail').exists()).toBe(true);

                // Expect the html to match the last testrun,
                // this is a little verbose though
                expect(wrapper.html()).toMatchSnapshot();
            });
        });
    });
});
