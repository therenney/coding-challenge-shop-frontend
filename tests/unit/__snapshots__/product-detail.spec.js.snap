// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetail Given there is product data When the ProductDetail component is rendered should render a view with the product details 1`] = `
<div>
  <div class="product-detail">
    <h2>
      Beverage 23, 6x 0.75 L<br><span class="uk-text-lead">â‚¬459.00</span></h2>
    <div uk-grid="" class="uk-grid-small uk-margin-bottom">
      <div class="uk-width-1-1 uk-width-1-2@s uk-width-1-3@m"><img data-src="https://dummyimage.com/600x400/eeeeee/2fd6ac.png&amp;text=Beverage+23,+6x0.75" width="100%" height="150" alt="" uk-img="" class="product-image"></div>
      <div class="uk-width-1-1 uk-width-1-2@s uk-width-2-3@m">
        <p>Lorem ipsum usu amet dicat nullam ea. Nec detracto lucilius democritum in, ne usu delenit offendit deterruisset. Recusabo iracundia molestiae ea pro, suas dicta nemore an cum, erat dolorum nonummy mel ea. Iisque labores liberavisse in mei, dico laoreet elaboraret nam et, iudico verterem platonem est an. Te usu paulo vidisse epicuri, facilis mentitum liberavisse vel ut, movet iriure invidunt ut quo. Ad melius mnesarchum scribentur eum, mel at mundi impetus utroque.</p>
      </div>
    </div>
    <div class="uk-column-1-3@m">
      <dl class="uk-description-list">
        <dt>Manufacturer</dt>
        <dd>
          Beverages Ltd.
        </dd>
        <dt>Packaging</dt>
        <dd>
          CA
        </dd>
        <dt>Base Product Packaging</dt>
        <dd>
          BO
        </dd>
        <dt>Base Product Unit</dt>
        <dd>
          LT
        </dd>
        <dt>Base Product Amount</dt>
        <dd>
          0.75
        </dd>
        <dt>Base Product Quantity</dt>
        <dd>
          6
        </dd>
      </dl>
    </div>
  </div>
</div>
`;
